File: trace/raytracer.cc
#include "trace/raytracer.h"

#include <glm/glm.hpp>
#include <glm/gtc/constants.hpp>

#include <optional>

#include "geometry/ray.h"
#include "geometry/triray.h"
#include "kdtree/kdtree.h"
#include "trace/camera.h"
#include "trace/fastrand.h"
#include "trace/light.h"
#include "trace/material.h"
#include "trace/mcsampling.h"
#include "trace/samplebuffer.h"
#include "trace/scene.h"

using geometry::Ray;
using geometry::TriRayIntersection;
using glm::vec3;
using std::optional;
using trace::FastRand;
using trace::Material;
using trace::Scene;
using trace::SphereLight;

namespace {
constexpr float EPSILON = 0.00001f;
}  // namespace

namespace trace {

vec3 Raytracer::LightContribution(const Scene& scene,
                                  const Material* material,
                                  const vec3& target,
                                  const vec3& offset,
                                  const vec3& wi,
                                  const vec3& n,
                                  const SphereLight& light) {
  vec3 source = light.GetCenter();
  vec3 direction = source - target;
  Ray shadow_ray{offset, direction};
  if (!scene.AnyIntersect(shadow_ray, 0.0f, 1.0f)) {
    vec3 wr = glm::normalize(direction);
    vec3 radiance = light.GetEmitted(target);
    return material->GetBrdf(wi, wr, n) * radiance * abs(dot(wr, n));
  }
  return glm::zero<vec3>();
}

vec3 Raytracer::EnvironmentContribution(const Ray&) const {
  return vec3(0.8f, 0.8f, 0.8f);
}

vec3 Raytracer::Trace(const Scene& scene, const Ray& ray) {
  optional<TriRayIntersection> intersection =
      scene.Intersect(ray, 0.0f, FLT_MAX);
  if (!intersection) return EnvironmentContribution(ray);

  vec3 wi = -ray.direction;
  vec3 point = intersection->get_position();
  vec3 n = intersection->get_normal();

  vec3 offset = point + EPSILON * n;

  const Material* material =
      static_cast<const Material*>(intersection->triangle->tag);

  vec3 radiance = glm::zero<vec3>();
  for (const SphereLight& light : scene.GetLights()) {
    radiance += LightContribution(scene, material, point, offset, wi, n, light);
  }

  return radiance;
}

void Raytracer::Render(const Scene& scene,
                       const Pinhole& pinhole,
                       SampleBuffer* buffer) {
  for (unsigned int y = 0; y < buffer->height(); ++y) {
    for (unsigned int x = 0; x < buffer->width(); ++x) {
      float sx = (static_cast<float>(x) + 0.5f) / buffer->width();
      float sy = (static_cast<float>(y) + 0.5f) / buffer->height();
      buffer->add(x, y, Trace(scene, pinhole.ray(sx, sy)));
    }
  }

  buffer->inc();
}

}  // namespace trace

File: tracing/src/raytracer.rs
use std::ops::RangeInclusive;

use crate::{
    image_buffer::ImageBuffer,
    material::{Material, OutgoingRay},
};
use geometry::ray::Ray;
use glam::{UVec2, Vec2, Vec3};
use kdtree::{intersection::KdIntersection, KdTree};
use scene::{camera::Pinhole, Scene};

pub struct Raytracer {
    pub scene: Scene,
    pub kdtree: KdTree,
    pub camera: Pinhole,
}

impl Raytracer {
    fn intersect(&self, ray: &Ray, t_range: RangeInclusive<f32>) -> Option<KdIntersection> {
        self.kdtree.intersect(ray, t_range)
    }

    fn intersect_any(&self, ray: &Ray, t_range: RangeInclusive<f32>) -> bool {
        self.intersect(ray, t_range).is_some()
    }

    fn trace_ray(&self, ray: &Ray) -> Vec3 {
        let intersection = self.intersect(ray, 0.0..=f32::MAX);
        if intersection.is_none() {
            return self.scene.environment;
        }
        let intersection = intersection.unwrap();
        let intersection_index = intersection.index;
        let intersection = intersection.intersection;

        let wi = -ray.direction;
        let point = ray.param(intersection.t);
        let n = self.scene.get_normal(intersection_index, &intersection);
        let uv = self.scene.get_texcoord(intersection_index, &intersection);

        // TODO: How to chose offset?
        let offset_point = point + 0.0001 * n;

        self.scene
            .lights
            .iter()
            .map(|light| {
                let direction = light.center - point;
                let shadow_ray = Ray::new(offset_point, direction);
                if self.intersect_any(&shadow_ray, 0.0..=1.0) {
                    return Vec3::ZERO;
                }
                let wo = direction.normalize();
                let radiance = light.emitted(point);
                let brdf = self
                    .scene
                    .get_material(intersection_index)
                    .brdf(&OutgoingRay { wi, n, wo, uv });
                brdf * radiance * wo.dot(n).abs()
            })
            .sum()
    }

    pub fn render(&self, buffer: &mut ImageBuffer) {
        let buffer_size = buffer.size.as_vec2();
        for y in 0..buffer.size.y {
            for x in 0..buffer.size.x {
                let pixel = UVec2::new(x, y);
                let pixel_center = Vec2::new(pixel.x as f32 + 0.5, pixel.y as f32 + 0.5);
                let scene_direction = pixel_center / buffer_size;
                let ray = self.camera.ray(scene_direction);
                buffer[pixel] += self.trace_ray(&ray);
            }
        }
    }

    pub fn camera(&self) -> &Pinhole {
        &self.camera
    }
}
